File: src/clicache/src/CqQuery.cpp
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

//#include "gf_includes.hpp"
#include "CqQuery.hpp"
#include "Query.hpp"
#include "CqAttributes.hpp"
#include "CqAttributesMutator.hpp"
#include "CqStatistics.hpp"
#include "ISelectResults.hpp"
#include "ResultSet.hpp"
#include "StructSet.hpp"
#include "ExceptionTypes.hpp"

using namespace System;

namespace Apache
{
  namespace Geode
  {
    namespace Client
    {
namespace Generic
    {

      generic<class TKey, class TResult>
      ICqResults<TResult>^ CqQuery<TKey, TResult>::ExecuteWithInitialResults()
      {
        return ExecuteWithInitialResults(DEFAULT_QUERY_RESPONSE_TIMEOUT);
      }

      generic<class TKey, class TResult>
      ICqResults<TResult>^ CqQuery<TKey, TResult>::ExecuteWithInitialResults(uint32_t timeout)
      {
        _GF_MG_EXCEPTION_TRY2/* due to auto replace */

          apache::geode::client::CqResultsPtr& nativeptr =
            NativePtr->executeWithInitialResults(timeout);
          if (nativeptr.ptr() == NULL) return nullptr;

          apache::geode::client::ResultSet* resultptr = dynamic_cast<apache::geode::client::ResultSet*>(
            nativeptr.ptr( ) );
          if ( resultptr == NULL )
          {
            apache::geode::client::StructSet* structptr = dynamic_cast<apache::geode::client::StructSet*>(
              nativeptr.ptr( ) );
            if ( structptr == NULL )
            {
              return nullptr;
            }
            return StructSet<TResult>::Create(structptr);
          }
          /*else
          {
            return ResultSet::Create(resultptr);
          }*/

          return nullptr;

        _GF_MG_EXCEPTION_CATCH_ALL2/* due to auto replace */
      }

      generic<class TKey, class TResult>
      void CqQuery<TKey, TResult>::Execute()
      {
        _GF_MG_EXCEPTION_TRY2/* due to auto replace */

          NativePtr->execute();

        _GF_MG_EXCEPTION_CATCH_ALL2/* due to auto replace */
      }

      generic<class TKey, class TResult>
      String^ CqQuery<TKey, TResult>::QueryString::get( )
      {
        return ManagedString::Get( NativePtr->getQueryString( ) );
      }

      generic<class TKey, class TResult>
      String^ CqQuery<TKey, TResult>::Name::get( )
      {
        return ManagedString::Get( NativePtr->getName( ) );
      }

      generic<class TKey, class TResult>
      Query<TResult>^ CqQuery<TKey, TResult>::GetQuery( )
      {
        return Query<TResult>::Create(NativePtr->getQuery().ptr());
      }

      generic<class TKey, class TResult>
      CqAttributes<TKey, TResult>^ CqQuery<TKey, TResult>::GetCqAttributes( )
      {
        return CqAttributes<TKey, TResult>::Create(NativePtr->getCqAttributes( ).ptr());
      }

      generic<class TKey, class TResult>
      CqAttributesMutator<TKey, TResult>^ CqQuery<TKey, TResult>::GetCqAttributesMutator( )
      {
        return CqAttributesMutator<TKey, TResult>::Create(NativePtr->getCqAttributesMutator().ptr());
      }

      generic<class TKey, class TResult>
      CqStatistics^ CqQuery<TKey, TResult>::GetStatistics( )
      {
        return CqStatistics::Create(NativePtr->getStatistics().ptr());
      }

      generic<class TKey, class TResult>
      CqStateType CqQuery<TKey, TResult>::GetState( )
      {
        apache::geode::client::CqState::StateType st =  NativePtr->getState( );
        CqStateType state;
        switch (st)
        {
          case apache::geode::client::CqState::STOPPED: {
            state = CqStateType::STOPPED;
            break;
          }
          case apache::geode::client::CqState::RUNNING: {
            state = CqStateType::RUNNING;
            break;
          }
          case apache::geode::client::CqState::CLOSED: {
            state = CqStateType::CLOSED;
            break;
          }
          case apache::geode::client::CqState::CLOSING: {
            state = CqStateType::CLOSING;
            break;
          }
          default: {
            state = CqStateType::INVALID;
            break;
          }
        }
        return state;
      }

      generic<class TKey, class TResult>
      void CqQuery<TKey, TResult>::Stop( )
      {
        _GF_MG_EXCEPTION_TRY2/* due to auto replace */

          NativePtr->stop( );

        _GF_MG_EXCEPTION_CATCH_ALL2/* due to auto replace */
      }

      generic<class TKey, class TResult>
      void CqQuery<TKey, TResult>::Close( )
      {
        _GF_MG_EXCEPTION_TRY2/* due to auto replace */

          NativePtr->close( );

        _GF_MG_EXCEPTION_CATCH_ALL2/* due to auto replace */
      }

      generic<class TKey, class TResult>
      bool CqQuery<TKey, TResult>::IsRunning( )
      {
        _GF_MG_EXCEPTION_TRY2/* due to auto replace */

          return NativePtr->isRunning( );

        _GF_MG_EXCEPTION_CATCH_ALL2/* due to auto replace */
      }

      generic<class TKey, class TResult>
      bool CqQuery<TKey, TResult>::IsStopped( )
      {
        _GF_MG_EXCEPTION_TRY2/* due to auto replace */

          return NativePtr->isStopped( );

        _GF_MG_EXCEPTION_CATCH_ALL2/* due to auto replace */
      }

      generic<class TKey, class TResult>
      bool CqQuery<TKey, TResult>::IsClosed( )
      {
        _GF_MG_EXCEPTION_TRY2/* due to auto replace */

          return NativePtr->isClosed( );

        _GF_MG_EXCEPTION_CATCH_ALL2/* due to auto replace */
      }

    }
  }
}
 } //namespace 

File: geode-core/src/main/java/org/apache/geode/cache/query/CqQuery.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more contributor license
 * agreements. See the NOTICE file distributed with this work for additional information regarding
 * copyright ownership. The ASF licenses this file to You under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance with the License. You may obtain a
 * copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */

package org.apache.geode.cache.query;


/**
 * Interface for continuous query. This provides methods for managing a CQ once it is created
 * through the QueryService. The methods allow you to retrieve CQ related information, operate on CQ
 * like execute, stop, close and get the state of the CQ.
 *
 * @since GemFire 5.5
 */

public interface CqQuery {

  /**
   * Get the original query string that was specified with CQ.
   *
   * @return String the query string associated with CQ.
   */
  String getQueryString();

  /**
   * Get the query object generated for this CQs query.
   *
   * @return Query query object for the query string.
   */
  Query getQuery();

  /**
   * Get the name of the CQ.
   *
   * @return the name of the CQ.
   */
  String getName();

  /**
   * Get statistics information for this CQ.
   *
   * @return CqStatistics CQ statistics object.
   */
  CqStatistics getStatistics();

  /**
   * Get CqAttributes for this CQ.
   *
   * @see CqAttributes
   * @return CqAttributes cqAttribute set with this CqQuery.
   */
  CqAttributes getCqAttributes();

  /**
   * Get CqAttributesMutator for this CQ.
   *
   * @see CqAttributesMutator
   * @return CqAttributesMutator.
   */
  CqAttributesMutator getCqAttributesMutator();

  /**
   * Starts executing the CQ or, if the CQ is in a stopped state, resumes execution. Gets the
   * current resultset for the CQ query. The CQ is executed on the primary server, and then
   * redundant servers as needed. If execution fails on all servers, a CqException is thrown. If the
   * query is complex and the data set is large, this execution may take a long time, which may
   * cause a socket read timeout in the client. To allow adequate time, you may need to set a longer
   * pool read-timeout in the client.
   *
   * @throws CqClosedException if this CqQuery is closed.
   * @throws RegionNotFoundException if the specified region in the query string is not found.
   * @throws IllegalStateException if the CqQuery is in the RUNNING state already.
   * @throws CqException if failed to execute and get initial results.
   * @return SelectResults resultset obtained by executing the query.
   */
  <E> CqResults<E> executeWithInitialResults()
      throws CqClosedException, RegionNotFoundException, CqException;

  /**
   * Start executing the CQ or if this CQ is stopped earlier, resumes execution of the CQ. The CQ is
   * executed on primary and redundant servers, if CQ execution fails on all the server then a
   * CqException is thrown.
   *
   * @throws CqClosedException if this CqQuery is closed.
   * @throws RegionNotFoundException if the specified region in the query string is not found.
   * @throws IllegalStateException if the CqQuery is in the RUNNING state already.
   * @throws CqException if failed to execute.
   */
  void execute() throws CqClosedException, RegionNotFoundException, CqException;

  /**
   * Stops this CqQuery without releasing resources. Puts the CqQuery into the STOPPED state. Can be
   * resumed by calling execute or executeWithInitialResults.
   *
   * @throws IllegalStateException if the CqQuery is in the STOPPED state already.
   * @throws CqClosedException if the CQ is CLOSED.
   */
  void stop() throws CqClosedException, CqException;

  /**
   * Get the state of the CQ in CqState object form. CqState supports methods like isClosed(),
   * isRunning(), isStopped().
   *
   * @see CqState
   * @return CqState state object of the CQ.
   */
  CqState getState();

  /**
   * Close the CQ and stop execution. Releases the resources associated with this CqQuery.
   *
   * @throws CqClosedException Further calls on this CqQuery instance except for getState() or
   *         getName().
   * @throws CqException - if failure during cleanup of CQ resources.
   */
  void close() throws CqClosedException, CqException;

  /**
   * This allows to check if the CQ is in running or active.
   *
   * @return boolean true if running, false otherwise
   */
  boolean isRunning();

  /**
   * This allows to check if the CQ is in stopped.
   *
   * @return boolean true if stopped, false otherwise
   */
  boolean isStopped();

  /**
   * This allows to check if the CQ is closed.
   *
   * @return boolean true if closed, false otherwise
   */
  boolean isClosed();

  /**
   * This allows to check if the CQ is durable.
   *
   * @return boolean true if durable, false otherwise
   * @since GemFire 5.5
   */
  boolean isDurable();

  /**
   * This allows to check if the CQ use option to suppress CQ create notification.
   *
   * @return boolean true if create is suppressed, false otherwise
   * @since Geode 14.0
   */
  boolean isCreateSuppressed();

  /**
   * This allows to check if the CQ use option to suppress CQ update notification.
   *
   * @return boolean true if update is suppressed, false otherwise
   * @since Geode 14.0
   */
  boolean isUpdateSuppressed();

  /**
   * This allows to check if the CQ use option to suppress CQ destroy notification.
   *
   * @return boolean true if destroy is suppressed, false otherwise
   * @since Geode 14.0
   */
  boolean isDestroySuppressed();

}
